src/character.cpp

```diff
@@ -7281,3 +7286,82 @@ void Character::use_fire( const int quantity )
+    //Average healthy bpm is 60-80. That's a simple imitation of mormal distribution.
+    //Must a better way to do that. Possibly this value should be generated with player creation.
+    int average_heartbeat = 70 + rng( -5, 5 ) + rng( -5, 5 );
+    //Chemical imbalance makes this less predictable. It's possible this range needs tweaking
+    if( has_trait( trait_CHEMIMBALANCE ) ) {
+        average_heartbeat += rng( -15, 15 );
+    }
+    //Quick also raises basic BPM
+    if( has_trait( trait_QUICK ) ) {
+        average_heartbeat *= 1.1;
+    }
+    //Badtemper makes youк BPM raise from anger
+    if( has_trait( trait_BADTEMPER ) ) {
+        average_heartbeat *= 1.1;
+    }
+    const float stamina_level = float( get_stamina() ) / float( get_stamina_max() );
+    float stamina_effect = 0;
+    if( stamina_level >= 0.9 ) {
+        stamina_effect = 0;
+    } else if( stamina_level >= 0.8 ) {
+        stamina_effect = 0.2;
+    } else if( stamina_level >= 0.6 ) {
+        stamina_effect = 0.5;
+    } else if( stamina_level >= 0.4 ) {
+        stamina_effect = 1;
+    } else if( stamina_level >= 0.2 ) {
+        stamina_effect = 1.5;
+    } else {
+        stamina_effect = 2;
+    }
+    //can triple heartrate
+    int heartbeat = average_heartbeat * ( 1 + stamina_effect );
+    const int stim_level = get_stim();
+    int stim_modifer = 0;
+    if( stim_level > 0 ) {
+        //that's asymptotical function that is equal to 1 at around 30 stim level
+        //and slows down all the time almost reaching 2.
+        //Tweaking x*x multiplier will accordingly change effect accumulation
+        stim_modifer = 2.1 - 2 / ( 1 + 0.001 * stim_level * stim_level );
+    }
+    heartbeat *= 1 + stim_modifer;
+    if( to_turns<int>( get_effect_dur( effect_cig ) ) > 0 ) {
+        //Nicotine-induced tachycardia
+        if( get_effect_dur( effect_cig ) > 10_minutes * ( addiction_level( ADD_CIG ) + 1 ) ) {
+            heartbeat *= 1.4;
+        } else {
+            heartbeat *= 1.1;
+        }
+    }
+    //health effect that can make things better or worse is applied in the end.
+    //Based on get_max_healthy that already has bmi factored
+    const int healthy = get_max_healthy();
+    float healthy_modifier = 0;
+    //a bit arbitary formula that can use some love
+    healthy_modifier = -0.05 * round( healthy / 20 );
+    heartbeat *= 1 + healthy_modifier;
+    //if BPM raised at least by 20% for a player with ADRENALINE, it adds 20% of avg to result
+    if( has_trait( trait_ADRENALINE ) && heartbeat > average_heartbeat * 1.2 ) {
+        heartbeat += average_heartbeat * 0.2;
+    }
+    //Add dependencies for COLDBLOOD?
+    //Add effects from addictions
+    //add morale effects, fear(?), medication effects
+    //A single clamp in the end should be enough
+    heartbeat = clamp( heartbeat, average_heartbeat, 250 );
+    //No heartbeat in omnicell
+    if( has_trait( trait_SLIMESPAWNER ) ) {
```

Comment:
Could add this to the beginning of function instead, so slime mutants dont have to calculate all the rest.

Reply:
Good point!